File: javascript-script-developer/assets/templates/file-processor.template.js
#!/usr/bin/env node
/**
 * Batch file processor.
 *
 * Walks a directory, processes each matching file with bounded concurrency,
 * writes results, reports progress.
 *
 * usage: <script-name> [options] <input-dir>
 */
import { parseArgs } from "node:util";
import { readdir, readFile, writeFile, mkdir, stat } from "node:fs/promises";
import { join, dirname, extname, relative } from "node:path";
import { fileURLToPath } from "node:url";
import process from "node:process";

let verbose = false;
const log = {
  debug: (...a) => verbose && process.stderr.write(`[debug] ${a.join(" ")}\n`),
  info: (...a) => process.stderr.write(`[info] ${a.join(" ")}\n`),
  warn: (...a) => process.stderr.write(`[warn] ${a.join(" ")}\n`),
  error: (...a) => process.stderr.write(`[error] ${a.join(" ")}\n`),
};

class UsageError extends Error {
  constructor(msg) {
    super(msg);
    this.exitCode = 2;
  }
}

// ---------- arg parsing ----------

function parseCliArgs() {
  const { values, positionals } = parseArgs({
    options: {
      output: { type: "string", short: "o", default: "./out" },
      pattern: { type: "string", short: "p", default: ".json" },
      concurrency: { type: "string", short: "c", default: "8" },
      verbose: { type: "boolean", short: "v" },
      help: { type: "boolean", short: "h" },
    },
    allowPositionals: true,
    strict: true,
  });

  if (values.help) {
    process.stderr.write(
      [
        "usage: <script-name> [options] <input-dir>",
        "",
        "options:",
        "  -o, --output <dir>     output directory (default: ./out)",
        "  -p, --pattern <ext>    file extension to match (default: .json)",
        "  -c, --concurrency <n>  parallel workers (default: 8)",
        "  -v, --verbose          enable debug logging",
        "  -h, --help             show this help",
        "",
      ].join("\n"),
    );
    process.exit(0);
  }

  if (positionals.length !== 1) {
    throw new UsageError(
      "expected exactly one positional argument: <input-dir>",
    );
  }

  const concurrency = Number(values.concurrency);
  if (!Number.isInteger(concurrency) || concurrency < 1) {
    throw new UsageError("--concurrency must be a positive integer");
  }

  return {
    inputDir: positionals[0],
    outputDir: values.output,
    pattern: values.pattern,
    concurrency,
    verbose: values.verbose ?? false,
  };
}

// ---------- file walking ----------

async function* walk(dir) {
  for (const entry of await readdir(dir, { withFileTypes: true })) {
    const full = join(dir, entry.name);
    if (entry.isDirectory()) {
      yield* walk(full);
    } else if (entry.isFile()) {
      yield full;
    }
  }
}

// ---------- bounded-concurrency pool ----------

async function pool(items, limit, fn) {
  const results = [];
  const executing = new Set();
  for (const item of items) {
    const p = Promise.resolve()
      .then(() => fn(item))
      .catch((err) => ({ error: err, item }));
    results.push(p);
    executing.add(p);
    p.finally(() => executing.delete(p));
    if (executing.size >= limit) {
      await Promise.race(executing);
    }
  }
  return Promise.all(results);
}

// ---------- per-file work ----------

async function processOne(srcPath, srcRoot, outRoot) {
  const rel = relative(srcRoot, srcPath);
  const outPath = join(outRoot, rel);
  await mkdir(dirname(outPath), { recursive: true });

  const contents = await readFile(srcPath, "utf8");

  // Replace with the actual transformation:
  const transformed = contents;

  // Atomic write
  const tmp = `${outPath}.tmp.${process.pid}`;
  await writeFile(tmp, transformed);
  const { rename } = await import("node:fs/promises");
  await rename(tmp, outPath);

  log.debug(`processed ${rel}`);
  return { rel, bytes: contents.length };
}

// ---------- main ----------

async function main() {
  const args = parseCliArgs();
  verbose = args.verbose;

  const inputStat = await stat(args.inputDir).catch(() => null);
  if (!inputStat?.isDirectory()) {
    throw new UsageError(`input is not a directory: ${args.inputDir}`);
  }

  // Graceful shutdown
  let interrupted = false;
  process.on("SIGINT", () => {
    interrupted = true;
    log.info("interrupted, finishing in-flight work");
  });

  // Collect matching files
  const allFiles = [];
  for await (const f of walk(args.inputDir)) {
    if (extname(f) === args.pattern) allFiles.push(f);
  }
  log.info(`found ${allFiles.length} files matching ${args.pattern}`);

  if (allFiles.length === 0) return 0;

  // Process with bounded concurrency
  const startedAt = Date.now();
  let done = 0;
  const results = await pool(allFiles, args.concurrency, async (f) => {
    if (interrupted) return { rel: f, skipped: true };
    const r = await processOne(f, args.inputDir, args.outputDir);
    done++;
    if (done % 50 === 0 || done === allFiles.length) {
      const pct = ((done / allFiles.length) * 100).toFixed(1);
      const elapsed = ((Date.now() - startedAt) / 1000).toFixed(1);
      process.stderr.write(
        `\r${done}/${allFiles.length} (${pct}%, ${elapsed}s)`,
      );
    }
    return r;
  });
  process.stderr.write("\n");

  const errors = results.filter((r) => r.error);
  const skipped = results.filter((r) => r.skipped);
  const ok = results.filter((r) => !r.error && !r.skipped);

  log.info(
    `ok: ${ok.length}, errors: ${errors.length}, skipped: ${skipped.length}`,
  );

  if (errors.length > 0) {
    for (const e of errors.slice(0, 5)) {
      log.error(`${e.item}: ${e.error.message}`);
    }
    return 1;
  }

  return interrupted ? 130 : 0;
}

// ---------- entrypoint ----------

const isEntry = process.argv[1] === fileURLToPath(import.meta.url);
if (isEntry) {
  main()
    .then((code) => process.exit(code ?? 0))
    .catch((err) => {
      if (err.exitCode != null) {
        log.error(err.message);
        process.exit(err.exitCode);
      }
      log.error("unexpected:", err.stack ?? err.message);
      process.exit(1);
    });
}

File: golang-script-developer/assets/templates/file-processor.template.go
// Package main is a batch file processor.
//
// Walks a directory, processes each matching file with bounded concurrency,
// writes results, reports progress.
//
// Usage: file-processor [options] <input-dir>
package main

import (
	"context"
	"errors"
	"flag"
	"fmt"
	"io/fs"
	"log/slog"
	"os"
	"os/signal"
	"path/filepath"
	"strings"
	"sync/atomic"
	"syscall"
	"time"

	"golang.org/x/sync/errgroup"
)

// ---------- exit codes ----------

const (
	exitOK        = 0
	exitGeneric   = 1
	exitUsage     = 2
	exitInterrupt = 130
)

type exitError struct {
	code int
	err  error
}

func (e *exitError) Error() string { return e.err.Error() }
func (e *exitError) Unwrap() error { return e.err }

func usageErr(format string, a ...any) error {
	return &exitError{code: exitUsage, err: fmt.Errorf(format, a...)}
}

// ---------- args ----------

type args struct {
	inputDir    string
	outputDir   string
	pattern     string
	concurrency int
	verbose     bool
}

func parseArgs() (args, error) {
	fs := flag.NewFlagSet("file-processor", flag.ContinueOnError)
	fs.SetOutput(os.Stderr)

	var a args
	fs.StringVar(&a.outputDir, "o", "./out", "output directory")
	fs.StringVar(&a.pattern, "p", ".json", "file extension to match")
	fs.IntVar(&a.concurrency, "c", 8, "parallel workers")
	fs.BoolVar(&a.verbose, "v", false, "enable debug logging")

	fs.Usage = func() {
		fmt.Fprintf(os.Stderr, "usage: file-processor [options] <input-dir>\n\n")
		fs.PrintDefaults()
	}

	if err := fs.Parse(os.Args[1:]); err != nil {
		return a, err
	}
	if fs.NArg() != 1 {
		return a, usageErr("expected exactly one positional argument: <input-dir>")
	}
	if a.concurrency < 1 {
		return a, usageErr("--concurrency must be positive, got %d", a.concurrency)
	}
	a.inputDir = fs.Arg(0)
	return a, nil
}

// ---------- per-file work ----------

func processOne(ctx context.Context, srcPath, srcRoot, outRoot string) error {
	if err := ctx.Err(); err != nil {
		return err
	}

	rel, err := filepath.Rel(srcRoot, srcPath)
	if err != nil {
		return fmt.Errorf("rel %q: %w", srcPath, err)
	}
	outPath := filepath.Join(outRoot, rel)

	if err := os.MkdirAll(filepath.Dir(outPath), 0o755); err != nil {
		return fmt.Errorf("mkdir: %w", err)
	}

	data, err := os.ReadFile(srcPath)
	if err != nil {
		return fmt.Errorf("read %q: %w", srcPath, err)
	}

	// Replace with the actual transformation:
	transformed := data

	// Atomic write
	tmp := outPath + ".tmp"
	if err := os.WriteFile(tmp, transformed, 0o644); err != nil {
		return fmt.Errorf("write tmp: %w", err)
	}
	if err := os.Rename(tmp, outPath); err != nil {
		_ = os.Remove(tmp)
		return fmt.Errorf("rename: %w", err)
	}

	slog.Debug("processed", "rel", rel)
	return nil
}

// ---------- main work ----------

func run(ctx context.Context, a args) error {
	info, err := os.Stat(a.inputDir)
	if err != nil {
		return fmt.Errorf("stat input: %w", err)
	}
	if !info.IsDir() {
		return usageErr("input is not a directory: %s", a.inputDir)
	}

	// Collect matching files first so we know the total
	var files []string
	err = filepath.WalkDir(a.inputDir, func(p string, d fs.DirEntry, err error) error {
		if err != nil { return err }
		if d.IsDir() { return nil }
		if strings.HasSuffix(p, a.pattern) {
			files = append(files, p)
		}
		return nil
	})
	if err != nil {
		return fmt.Errorf("walk: %w", err)
	}
	slog.Info("found files", "count", len(files), "pattern", a.pattern)
	if len(files) == 0 {
		return nil
	}

	// Process with bounded concurrency
	g, gctx := errgroup.WithContext(ctx)
	g.SetLimit(a.concurrency)

	var done atomic.Int64
	startedAt := time.Now()

	for _, f := range files {
		f := f // capture for closure (unnecessary in Go 1.22+, kept for clarity)
		g.Go(func() error {
			if err := processOne(gctx, f, a.inputDir, a.outputDir); err != nil {
				return fmt.Errorf("%s: %w", f, err)
			}
			n := done.Add(1)
			if n%50 == 0 || n == int64(len(files)) {
				pct := float64(n) / float64(len(files)) * 100
				fmt.Fprintf(os.Stderr, "\r%d/%d (%.1f%%, %.1fs)",
					n, len(files), pct, time.Since(startedAt).Seconds())
			}
			return nil
		})
	}

	if err := g.Wait(); err != nil {
		fmt.Fprintln(os.Stderr) // newline after the progress carriage return
		return err
	}
	fmt.Fprintln(os.Stderr)
	slog.Info("done", "ok", done.Load())
	return nil
}

// ---------- main ----------

func main() {
	a, err := parseArgs()
	if err != nil {
		if errors.Is(err, flag.ErrHelp) { os.Exit(exitOK) }
		fmt.Fprintln(os.Stderr, "error:", err)
		var ee *exitError
		if errors.As(err, &ee) { os.Exit(ee.code) }
		os.Exit(exitUsage)
	}

	level := slog.LevelInfo
	if a.verbose { level = slog.LevelDebug }
	slog.SetDefault(slog.New(slog.NewTextHandler(os.Stderr,
		&slog.HandlerOptions{Level: level})))

	ctx, stop := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
	defer stop()

	if err := run(ctx, a); err != nil {
		var ee *exitError
		if errors.As(err, &ee) {
			slog.Error(ee.Error())
			os.Exit(ee.code)
		}
		if errors.Is(err, context.Canceled) {
			slog.Info("interrupted")
			os.Exit(exitInterrupt)
		}
		slog.Error("unexpected", "err", err)
		os.Exit(exitGeneric)
	}
}
